File: core/src/nl/t64/game/rpg/components/conversation/ConversationContainer.java
package nl.t64.game.rpg.components.conversation;

import com.badlogic.gdx.Gdx;
import nl.t64.game.rpg.Utils;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;


public class ConversationContainer {

    private static final String CONVERSATION_CONFIGS = "configs/conversations/";
    private static final String FILE_LIST = CONVERSATION_CONFIGS + "_files.txt";

    private final Map<String, ConversationGraph> conversations;

    public ConversationContainer() {
        this.conversations = new HashMap<>();
        this.loadConversations();
    }

    public ConversationGraph getConversationById(String conversationId) {
        return conversations.get(conversationId);
    }

    public PhraseIdContainer createPhraseIdContainer() {
        var phraseIdContainer = new PhraseIdContainer();
        conversations.forEach((id, graph) -> phraseIdContainer.setPhraseId(id, graph.currentPhraseId));
        return phraseIdContainer;
    }

    public void setCurrentPhraseIds(PhraseIdContainer phraseIdContainer) {
        conversations.forEach((id, graph) -> graph.currentPhraseId = phraseIdContainer.getPhraseId(id));
    }

    private void loadConversations() {
        String[] configFiles = Gdx.files.local(FILE_LIST).readString().split(System.lineSeparator());
        Arrays.stream(configFiles)
              .map(filePath -> Gdx.files.local(CONVERSATION_CONFIGS + filePath).readString())
              .map(json -> Utils.readValue(json, ConversationGraph.class))
              .forEach(conversations::putAll);
    }

}

File: core/src/nl/t64/game/rpg/components/conversation/ConversationContainer.kt
package nl.t64.game.rpg.components.conversation

import com.badlogic.gdx.Gdx
import nl.t64.game.rpg.Utils


private const val CONVERSATION_CONFIGS = "configs/conversations/"
private const val FILE_LIST = CONVERSATION_CONFIGS + "_files.txt"

class ConversationContainer {

    private val conversations: Map<String, ConversationGraph> = Gdx.files.internal(FILE_LIST).readString()
        .split(System.lineSeparator())
        .map { Gdx.files.internal(CONVERSATION_CONFIGS + it).readString() }
        .map { Utils.readValue<ConversationGraph>(it) }
        .flatMap { it.toList() }
        .toMap()

    fun getConversationById(conversationId: String): ConversationGraph {
        return conversations[conversationId]!!
    }

    fun createPhraseIdContainer(): PhraseIdContainer {
        val phraseIdContainer = PhraseIdContainer()
        conversations.forEach { phraseIdContainer.setPhraseId(it.key, it.value.currentPhraseId) }
        return phraseIdContainer
    }

    fun setCurrentPhraseIds(phraseIdContainer: PhraseIdContainer) {
        conversations.forEach { it.value.currentPhraseId = phraseIdContainer.getPhraseId(it.key) }
    }

}
